cli/cli-extensions.md

```diff
@@ -0,0 +1,195 @@
+# CLI Extensions
+
+## Overview
+
+CLI extensions are separate executables that appear as part of the `temporal` CLI. This is similar to
+[`kubectl` plugins](https://kubernetes.io/docs/tasks/extend-kubectl/kubectl-plugins/) and `git` extensions/commands.
+
+## Usage Examples
+
+### Custom Authenticator
+
+A custom extension can allow
+
+    temporal mycompany login
+
+To set some environment config settings on disk so that successive calls to
+
+    temporal workflow start --type foo --task-queue some-tq --workflow-id foo-id
+
+automatically work.
+
+### Shortcuts for Common Tasks
+
+A custom extension can add something like
+
+    temporal mycompany do-company-thing
+
+And that can do anything programmatically and also get the benefit of having access to the Temporal client. So it could
+access some company resource (e.g. a database) and send an update to a workflow with some value.
+
+### Contributions
+
+Custom extensions allow an ecosystem of CLI extensions to form that do not require Temporal oversight. For example
+
+    temporal workflow show-diagram --workflow-id foo-id
+
+Could be an extension that downloads the workflow history and makes a nice visualization of it.
+
+### Cloud CLI
+
+A Temporal Cloud extension will be available (and likely released alongside the traditional CLI) that will allow
+cloud-specific operations like:
+
+    temporal cloud namespace list
+
+This can share authentication with the CLI in many ways. It can even support login, so the following two commands could
+work well together:
+
+    temporal cloud login --profile cloud
+    temporal workflow start --profile cloud --type foo --task-queue some-tq --workflow-id foo-id
+
+These are just usage examples for the purposes of explaining extensions, this may not be what `temporal cloud` extension
+looks like when built.
+
+## Runtime Behavior
+
+### Lookup
+
+Extensions are any executables on the `PATH` with `temporal-` prefix. Extension executable names use `-` for each
+subcommand and `_` for each dash in the command (but still can be called with an underscore). This is similar to
+`kubectl` behavior.
+
+The `PATH` lookup is most-specific to least-specific. For example, running `temporal foo bar-baz qux` will try to match
+the following in order:
+
+* `temporal-foo-bar_baz-qux`
+* `temporal-foo-bar_baz`
+* `temporal-foo`
+
+And the first one found will be executed with all arguments. And since underscores and dashes are the same,
+`temporal-foo-bar_baz` executable is called for both `temporal foo bar-baz qux` and `temporal foo bar_baz qux` though
+help text only shows the former.
+
+Built-in commands cannot be overridden by extensions, but subcommands can be added to existing commands.
+
```

Comment:
It might make sense to reserve some names as unavailable for extensions, for example `temporal nexus`, `temporal config` etc?

Beyond that, are you going to impose any restrictions on valid names? (File systems are very permissive) We obviously want to support user's local language; I think that typically means allowing non-utf8 even (?).

Reply:
> It might make sense to reserve some names as unavailable for extensions

Yes we can reserve some, though we won't predict them all. I will make a blanket statement like "Built-in commands always take precedent so if a newer Temporal CLI has a built-in command that once was an extension, the command will be chosen and the extension name will have to be changed". This is a normal state of affairs for CLI extensions, and in many ways it can be a good thing because some extensions could be contributions upstreamed to become built-in.